src/System.Linq/tests/GroupJoinTests.cs

```diff
@@ -0,0 +1,419 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using Xunit;
+
+namespace System.Linq.Tests
+{
+    public class GroupJoinTests
+    {
+        private class AnagramEqualityComparer : IEqualityComparer<string>
+        {
+            public bool Equals(string x, string y)
+            {
+                if (ReferenceEquals(x, y)) return true;
+                if (x == null | y == null) return false;
+                int length = x.Length;
+                if (length != y.Length) return false;
+                using (var en = x.OrderBy(i => i).GetEnumerator())
+                {
+                    foreach (char c in y.OrderBy(i => i))
+                    {
+                        en.MoveNext();
+                        if (c != en.Current) return false;
+                    }
+                }
+                return true;
+            }
+
+            public int GetHashCode(string obj)
+            {
+                int hash = 0;
+                foreach (char c in obj)
+                    hash ^= (int)c;
+                return hash;
+            }
+        }
+
+        public struct CustomerRec
+        {
+            public string name;
+            public int? custID;
+        }
+
+        public struct OrderRec
+        {
+            public int? orderID;
+            public int? custID;
+            public int? total;
+        }
+
+        public struct AnagramRec
+        {
+            public string name;
+            public int? orderID;
+            public int? total;
+        }
+
+        public struct JoinRec : IEquatable<JoinRec>
+        {
+            public string name;
+            public int?[] orderID;
+            public int?[] total;
+            
+            public override int GetHashCode()
+            {
+                // Not great, but it'll serve.
+                return name.GetHashCode() ^ orderID.Length ^ (total.Length * 31);
+            }
+            
+            public bool Equals(JoinRec other)
+            {
+                if(!string.Equals(name, other.name))
```

Comment:
Nit: throughout the code, please ensure there are spaces between keywords like ```if``` and ```for``` and subsequent parens.